crates/driver/src/domain/competition/mod.rs

```diff
@@ -95,36 +97,100 @@ impl Competition {
         .await;
 
         // Filter out solutions that failed verification.
-        let settlements = settlements.into_iter().filter_map(|(id, result)| {
-            result
-                .tap_err(|err| {
-                    tracing::info!(?err, ?id, "discarding solution: failed verification")
-                })
-                .ok()
-        });
-
-        // Score the solutions.
-        let scores = join_all(settlements.map(|settlement| async move {
-            tracing::trace!(id = ?settlement.solution.id, "scoring");
-            (settlement.score(&self.eth, auction).await, settlement)
-        }))
-        .await;
-
-        // Filter out solutions which failed scoring.
-        let scores = scores
+        let mut settlements = settlements
             .into_iter()
-            .filter_map(|(result, settlement)| {
+            .filter_map(|(id, result)| {
                 result
                     .tap_err(|err| {
-                        tracing::info!(?err, id = ?settlement.solution.id, "discarding solution: failed scoring")
+                        tracing::info!(?err, ?id, "discarding solution: failed verification")
                     })
                     .ok()
-                    .map(|score| (score, settlement))
-            });
+            })
+            .collect_vec();
+
+        // TODO(#1483): parallelize this
+
+        // Merge the settlements in random order.
+        settlements.shuffle(&mut rand::thread_rng());
+        // The merged settlements in their final form.
+        let mut results = Vec::new();
+        while !settlements.is_empty() {
+            let settlement = settlements.pop().unwrap();
+            // Has [`settlement`] been merged into another settlement?
+            let mut merged = false;
+            // Try to merge [`settlement`] into some other settlement.
+            for other in settlements.iter_mut() {
+                match other
+                    .merge(
+                        &settlement,
+                        self.eth.contracts().settlement(),
+                        &self.eth,
+                        &self.simulator,
+                    )
+                    .await
+                {
+                    Ok(m) => {
+                        tracing::debug!(
+                            settlement_1 = ?settlement.solutions(),
+                            settlement_2 = ?other.solutions(),
+                            "merged settlements"
+                        );
+                        *other = m;
+                        merged = true;
+                        break;
```

Comment:
Always merging a candidate settlement into some other settlement at most once would mean that we don't look at a lot of possible candidates.
This is probably fine for now when we want to get **something** that works but this might need to be improved in the future.

Reply:
As mentioned in the PR description, this is noted in https://github.com/cowprotocol/services/issues/1480.